scapy/modules/packet_viewer/viewer.py

```diff
@@ -0,0 +1,202 @@
+                        If a basecls is provided. The Viewer will automatically
+                        read basecls specific column configuration from
+                        `conf.contribs["packet_viewer_columns"]`.
+        :param views: Custom or additional views.
+        :param kwargs_for_sniff: Arguments for sniff, if source is a socket.
+        """
+        self.palette = [
+            ("header", "black", "dark green"),
+            ("row_focused", "black", "dark cyan"),
+            ("green", "black", "dark green"),
+            ("red", "white", "dark red"),
+            ("default_bold", "bold", ""),
+        ]
+
+        if views is None:
+            self.views = [ShowView]
+            try:
+                self.views += conf.contribs["packet_viewer_custom_views"]
+            except KeyError:
+                pass
+
+        for view in self.views:
+            self.palette += getattr(view, "palette", [])
+
+        self.source = source
+        self.kwargs_for_sniff = kwargs_for_sniff
+        self.formatter = RowFormatter(columns, basecls)
+        self.main_window = None   # type: Optional[MainWindow]
+        self.loop = None          # type: Optional[MainLoop]
+
+    def _connect_signals(self):
+        # type: () -> None
+        """
+        Internal function to connect signals from MainWindow to PopUps
+        """
+        if self.main_window is None:
+            return
+
+        connect_signal(
+            self.main_window, "question_popup",
+            lambda _, msg, cb: show_question_pop_up(self.loop, msg, cb))
+
+        connect_signal(
+            self.main_window, "input_popup",
+            lambda _, caption, initial, button_text, callback: show_input_pop_up(self.loop, caption, initial, button_text, callback))  # noqa: E501
+
+        connect_signal(
+            self.main_window, "info_popup",
+            lambda _, info: show_info_pop_up(self.loop, info))
+
+        connect_signal(
+            self.main_window.packet_view, "packets_modified",
+            self._update_screen)
+
+    def run(self):
+        # type: () -> Tuple[PacketList, PacketList]
+        """
+        Start Viewer
+        :return: Tuple of two PacketLists. First list contains all selected
+                 Packets. Second list contains all Packets
+        """
+        cf = conf.color_theme
+        conf.color_theme = BlackAndWhite()
+        try:
+            self.main_window = MainWindow(self.source, self.formatter,
+                                          self.views, **self.kwargs_for_sniff)
+
+            self.loop = MainLoop(self.main_window, palette=self.palette,
+                                 screen=ScreenWSL())
+
+            self._connect_signals()
+            self.loop.run()
+            return self.main_window.selected_packets, \
+                self.main_window.all_packets
+        except Exception as e:
+            print(e)
+            raise e
+        finally:
+            conf.color_theme = cf
```

Comment:
Warum kommt das in finally und nicht vor das return in Zeile 163?
Hat das noch einen Effekt, wenn eine Exception geschmissen wird?

Reply:
In case of an exception, the code would not be executed. Through finally, this code always get's executed even if an exception is raised before. Otherwise the color sheme of the interactive scapy shell is messed up.